### PR TITLE
Added additional telemetry to VsTest task

### PR DESCRIPTION
**Pull request by ShreyasRmsft**

### CODE REVIEW

File: Tasks/VsTest/vstest.ts [hunk 1]
```diff
@@ -122,6 +127,7 @@ function getVstestArguments(settingsFile: string, tiaEnabled: boolean): string[]
             if (!tl.exist(settingsFile)) {
                 // because this is filepath input build puts default path in the input. To avoid that we are checking this.
                 tl.setResult(tl.TaskResult.Failed, tl.loc('InvalidSettingsFile', settingsFile));
+                publishErrorToCi(constants.errorCodes.invalidSettingsFile, "Invalid settings file");
```
Comment (cltshivash) [inline Tasks/VsTest/vstest.ts]: >publishErrorToCi [](start = 16, length = 16)

Can this precede the setResult ?

**Comment by cltshivash** [comment]
>                                                                         defer.resolve(tl.TaskResult.Failed);

Any reason to ignore this ..? and couple of other above..?

---
Refers to: Tasks/VsTest/vstest.ts:627 in 4bd6fb1. [](commit_id = 4bd6fb131903479eaf5fa0664dd04495e21e0968, deletion_comment = False)

**Comment by cltshivash** [comment]
:clock1:

File: Tasks/VsTest/constants.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+export module errorCodes {
```
Comment (nigurr) [inline Tasks/VsTest/constants.ts]: Should start with "Capital letter". Please install tslint plugin in vscode and it should warn these

File: Tasks/VsTest/constants.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    export const publishResults = 'publishResults';
+    export const invokeVsTest = 'invokeVsTest';
+    export const runTestsLocally = 'runTestsLocally';
+    export const invalidSettingsFile = 'invalidSettingsFile';
+    export const executeVstest = 'executeVstest';
```
Comment (nigurr) [inline Tasks/VsTest/constants.ts]: keep the actual string value in lowercase

File: Tasks/VsTest/vstest.ts [hunk 4]
```diff
@@ -28,6 +29,7 @@ const systemDefaultWorkingDirectory = tl.getVariable('System.DefaultWorkingDirec
 const workingDirectory = systemDefaultWorkingDirectory;
 let testAssemblyFiles = undefined;
 let resultsDirectory = null;
+const taskProps = { errorCode: '', errorMessage: ''};
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: use "result" for publishing the error message "resultCode" for code. we are already using this property

File: Tasks/VsTest/vstest.ts [hunk 5]
```diff
@@ -10,6 +10,7 @@ import * as utils from './helpers';
 import * as outStream from './outputstream';
 import * as ci from './cieventlogger';
 import * as testselectorinvoker from './testselectorinvoker';
+import * as constants from './constants';
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: import errorCodes directly.

File: Tasks/VsTest/vstest.ts [hunk 6]
```diff
@@ -879,4 +894,10 @@ function responseContainsNoTests(filePath: string): Q.Promise<boolean> {
             return false;
         }
     });
+}
+
+function publishErrorToCi(errorCode, errorMessage) {
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: initialize taskProps in local. not required global in this case. and use strongly typed properties

File: Tasks/VsTest/vstest.ts (hunk 1 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/vstest.ts]: my bad

---
In reply to: [136021659](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136021659) [](ancestors = 136021659)

**Comment by ShreyasRmsft** [comment]
>                                                                         defer.resolve(tl.TaskResult.Failed);

Will be adding ci data here.

---
In reply to: [325940897](https://github.com/Microsoft/vsts-tasks/pull/5156#issuecomment-325940897) [](ancestors = 325940897)

---
Refers to: Tasks/VsTest/vstest.ts:627 in 4bd6fb1. [](commit_id = 4bd6fb131903479eaf5fa0664dd04495e21e0968, deletion_comment = False)

File: Tasks/VsTest/constants.ts (hunk 3 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/constants.ts]: I installed. It didn't recommend i make it Capital.

---
In reply to: [136024318](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136024318) [](ancestors = 136024318)

File: Tasks/VsTest/constants.ts (hunk 2 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/constants.ts]: tslint didnt make that suggestion

---
In reply to: [136024181](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136024181) [](ancestors = 136024181)

File: Tasks/VsTest/constants.ts (hunk 3 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/constants.ts]: done

---
In reply to: [136024318](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136024318) [](ancestors = 136024318)

File: Tasks/VsTest/vstest.ts (hunk 5 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/vstest.ts]: done

---
In reply to: [136024846](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136024846) [](ancestors = 136024846)

File: Tasks/VsTest/vstest.ts (hunk 6 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/vstest.ts]: strongly typed done

File: Tasks/VsTest/constants.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+export module areaCodes {
```
Comment (nigurr) [inline Tasks/VsTest/constants.ts]: It should start with Capital Letter

File: Tasks/VsTest/constants.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    export const publishResults = 'publishresults';
```
Comment (nigurr) [inline Tasks/VsTest/constants.ts]: const always FULL CAPITALS

File: Tasks/VsTest/vstest.ts [hunk 9]
```diff
@@ -28,6 +29,7 @@ const systemDefaultWorkingDirectory = tl.getVariable('System.DefaultWorkingDirec
 const workingDirectory = systemDefaultWorkingDirectory;
 let testAssemblyFiles = undefined;
 let resultsDirectory = null;
+const taskProps = { areaCode: '', message: '', tracePoint: '0', isUserError: 'false'};
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: result

Comment (nigurr) [reply Tasks/VsTest/vstest.ts]: properties small letters

Comment (nigurr) [reply Tasks/VsTest/vstest.ts]: discuss with @RanjanarMS. i dont like duplicated data

File: Tasks/VsTest/vstest.ts [hunk 10]
```diff
@@ -65,17 +67,20 @@ export function startTest() {
...
             })
             .fail(function (err) {
+                publishEventToCi(areaCodes.invokeVsTest, err.message);
                 deleteVstestDiagFile();
                 console.log('##vso[task.logissue type=error;code=' + err + ';TaskName=VSTest]');
-                throw err;
+                throw err;  // What is the point of this throw? Where is it getting caught?
             });
     } catch (error) {
+        publishEventToCi(areaCodes.runTestsLocally, error.message);
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: JSON.stringfy

File: Tasks/VsTest/vstest.ts [hunk 11]
```diff
@@ -121,6 +126,7 @@ function getVstestArguments(settingsFile: string, tiaEnabled: boolean): string[]
         } else {
             if (!tl.exist(settingsFile)) {
                 // because this is filepath input build puts default path in the input. To avoid that we are checking this.
+                publishEventToCi(areaCodes.invalidSettingsFile, 'Invalid settings file');
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: InvalidSettingsFile

File: Tasks/VsTest/vstest.ts [hunk 12]
```diff
@@ -522,15 +530,18 @@ function runVStest(testResultsDirectory: string, settingsFile: string, vsVersion
                                             uploadTestResults(testResultsDirectory)
                                                 .then(function (code) {
                                                     if (!isNaN(+code) && +code !== 0) {
+                                                        publishEventToCi(areaCodes.uploadTestResults, 'Upload test results returned ' + code);
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: keep all these strings constants.ts

File: Tasks/VsTest/vstest.ts [hunk 13]
```diff
@@ -874,9 +913,14 @@ function responseContainsNoTests(filePath: string): Q.Promise<boolean> {
...
             return true;
-        }
-        else {
+        } else {
             return false;
         }
     });
+}
+
+function publishEventToCi(areaCode : string, message: string) {
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: const taskProps = { areaCode: '', message: '', tracePoint: '0', isUserError: 'false'};

File: Tasks/VsTest/vstest.ts (hunk 10 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/vstest.ts]: done

---
In reply to: [136252940](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136252940) [](ancestors = 136252940)

File: Tasks/VsTest/vstest.ts (hunk 9 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/vstest.ts]: done

---
In reply to: [136252600](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136252600) [](ancestors = 136252600)

Comment (ShreyasRmsft) [reply Tasks/VsTest/vstest.ts]: done

---
In reply to: [136252471](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136252471) [](ancestors = 136252471)

File: Tasks/VsTest/vstest.ts (hunk 6 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/vstest.ts]: done

---
In reply to: [136025114](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136025114) [](ancestors = 136025114)

File: Tasks/VsTest/vstest.ts (hunk 13 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/vstest.ts]: done

---
In reply to: [136253659](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136253659) [](ancestors = 136253659)

File: Tasks/VsTest/constants.ts (hunk 8 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/constants.ts]: done

---
In reply to: [136252235](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136252235) [](ancestors = 136252235)

File: Tasks/VsTest/vstest.ts (hunk 10 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/vstest.ts]: This doesn't get the message out. Reverting to error.message

---
In reply to: [136255079](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136255079) [](ancestors = 136255079,136252940)

File: Tasks/VsTest/vstest.ts [hunk 14]
```diff
@@ -65,17 +66,20 @@ export function startTest() {
...
                 } catch (error) {
+                    publishEventToCi(AreaCodes.PUBLISHRESULTS, error.message, '0');
                     deleteVstestDiagFile();
                     console.log('##vso[task.logissue type=error;code=' + error + ';TaskName=VSTest]');
-                    throw error;
+                    throw error; // This throw should be removed
                 }
             })
             .fail(function (err) {
+                publishEventToCi(AreaCodes.INVOKEVSTEST, err.message, '1');
```
Comment (cltshivash) [inline Tasks/VsTest/vstest.ts]: nitpick: start with 1001..

File: Tasks/VsTest/vstest.ts [hunk 15]
```diff
@@ -874,9 +934,17 @@ function responseContainsNoTests(filePath: string): Q.Promise<boolean> {
...
             return true;
-        }
-        else {
+        } else {
             return false;
         }
     });
+}
+
+function publishEventToCi(areaCode : string, message: string, tracepoint: string, isusererror: string = 'false') {
```
Comment (cltshivash) [inline Tasks/VsTest/vstest.ts]: can this be a int (inline with service side)

Comment (ShreyasRmsft) [reply Tasks/VsTest/vstest.ts]: Had started with ints. Nitin had initially asked that i make it a string for some reason. Will revert to integers starting from 1001.

---
In reply to: [136512464](https://github.com/Microsoft/vsts-tasks/pull/5156#discussion_r136512464) [](ancestors = 136512464)

File: Tasks/VsTest/vstest.ts [hunk 16]
```diff
@@ -874,9 +934,17 @@ function responseContainsNoTests(filePath: string): Q.Promise<boolean> {
...
             return true;
-        }
-        else {
+        } else {
             return false;
         }
     });
+}
+
+function publishEventToCi(areaCode : string, message: string, tracepoint: number, isusererror: string = 'false') {
```
Comment (RanjanarMS) [inline Tasks/VsTest/vstest.ts]: isUserError

**Pull request merged**